fix(api): dont change curwin for nvim_win_set_height

folke opened the pull request:
Below is the code that calculates the minimum height.

When using `nvim_win_set_height`, `curwin` is temporarily changed to the window we're changing. This means that the code below will always set `height=2` when a winbar is active on the window.

This PR fixes that.

```
// Always keep current window at least one line high, even when 'winminheight' is zero.
  // Keep window at least two lines high if 'winbar' is enabled.
```

https://github.com/neovim/neovim/blob/2bdef6dd2a7572602aeb2efec76812769bcee246/src/nvim/window.c#L5788

folke committed:
fix(api): dont change curwin for nvim_win_set_height

folke committed:
test: added tests for set_height with winminheight=0 and a winbar

zeertzjq commented:
Will this cause other problems?

zeertzjq commented:
It's also a bit inconsistent that this change is not done for `nvim_win_set_width`

folke committed:
fix(api): dont change curwin for nvim_win_set_width

folke commented:
@zeertzjq good point on set_width. Changed that one as well. That wasn't causing issues though.

The change is in line with how heights are supposed to work. see `h winminheight`

Without this PR, I had to do:
1. `nvim_win_set_height = 1` (result would be height=2)
2. `win_move_statusline(win, -1)`

folke commented:
Don't merge this yet. I think there's an issue with the autocmds that trigger based on `nvim_win_set_height`. They probably expect the `curwin` to be the one we're changing.

folke commented:
Fixed the issue with splitkeep.

What happened, is I have a buffer open, and an autocmd changed the height of another window.

That triggered the `win_fix_scroll` function with `resize=true`.

That in turn executes `win_fix_cursor(true)`, which will update the cursor in the current window.

The current window has not been resized, so that's not something that should run.

Fixed by executing `win_fix_scroll` with `resize = curwin == win`.

@luukvbaal would be great if you could confirm my change related to splitkeep is correct :)

luukvbaal commented:
Yes that looks correct, is this something that could have happened without this change?
If so, the splitkeep fix might need to be submitted to vim first.

folke commented:
Looking at all the places where `win_fix_scroll` and `win_fix_cursor` are called, I think it's fine.

luukvbaal commented:
> That in turn executes `win_fix_cursor(true)`, which will update the cursor in the current window.

I do wonder why this happened though, do you have a repro for the issue you encountered?
If the height of `curwin` didn't change, I would expect that `win_fix_cursor()` doesn't change the cursor position either.

But you're right that `win_fix_cursor()` shouldn't (have to) be called when it's height didn't change. Getting rid of the `resize` argument and determining whether `curwin` height changed in `win_fix_scroll()` directly might be better.

folke commented:
I had two windows left and right.

My cursor was in the window on the right.

I hit `gd` on a function on the right, that opens a new buffer and goes to the location.

But I also had an autocmd on `FileType` that changed the height of the window on the left.

That autocmd triggered fix cursor in the current window, which was the window on the right.

The result was that instead of jumping to the correct definition in the new buffer that was loaded on the right, splitkeep ***fixed** my scroll position and so it was no longer where I should be.

luukvbaal commented on src/nvim/window.c:
```diff
@@ -5812,7 +5812,8 @@ void win_setheight_win(int height, win_T *win)
     if (*p_spk != 'c') {
-      win_fix_scroll(true);
+      // cursor only needs to be fixed if it's in the current window
+      win_fix_scroll(win == curwin);
```
This will conflict with #23884. I'm also not so sure it's correct anymore.
It had been a while since I worked on 'splitkeep', and the way the `resize` argument is used is confusing. But even if `win != curwin` here, `curwin`'s height might have changed if it is in the same frame as `win`, in which case the cursor _would_ need to be adjusted.

luukvbaal replied:
It would be good to grasp the problem you're trying to fix with an actual reproducible `test.lua` but perhaps you could work around the behavior by deferring the jump to definition instead?

folke replied:
In my particular test, the current window's height didn't change. I jumped to a new buffer loaded in the same window to a certiain location (`vim.lsp.buf.jump_to_location`). After that an autocmd ran that adjusted the height of an unrelated window, but the fix cursor code then ***fixed*** the cursor position of the current window/buffer, which is not correct. Result being that I'm not at the location I jumped to.

What would be a better fix for this? I'm not that fmailiar with the splitkeep code.

folke replied:
It seemed to adjust the cursor of the current window to the height of the window I was changing the height for

luukvbaal replied:
Wouldn't that have happened _before_ your change to `nvim_win_set_height()`, but not anymore?

folke replied:
Before my change, the current window would indeed have been set to the window that's changing height, so I guess in that case it's indeed needed to fix the cursor?

But with my new change, the current window isn't always the window we're changing.

folke replied:
I'll try yo create a minimal config to trigger the issue

folke commented:
@luukvbaal I have a (not so short) minimal init.lua.

I can actually reproduce this without this PR, so my changes probably just enlarge the issue somehow.

```lua
vim.o.splitkeep = "screen"
vim.o.number = true

-- create test file
local lines = {}
for i = 1, 100, 1 do
	lines[#lines + 1] = i .. ""
end
local fd = assert(io.open("./foo.bar", "w+"))
fd:write(table.concat(lines, "\n"))
fd:close()

local win = vim.api.nvim_get_current_win()

-- create second window
vim.cmd([[vnew]])
local win2 = vim.api.nvim_get_current_win()
vim.api.nvim_buf_set_lines(0, 0, -1, false, { "Second Window" })

-- make the previous window active
vim.api.nvim_set_current_win(win)

vim.api.nvim_create_autocmd({ "WinResized" }, {
	callback = function()
		-- set to same height
		local height = vim.api.nvim_win_get_height(win2)
		-- commenting the line below, will make the issue go away
		vim.api.nvim_win_set_height(win2, height)
	end,
})

-- simulates vim.lsp.buf.jump_to_location()
local function go(line)
	local win = vim.api.nvim_get_current_win()
	local bufnr = vim.fn.bufadd("./foo.bar")
	vim.api.nvim_win_set_buf(win, bufnr)
	-- vim.api.nvim_set_current_win(win)
	vim.api.nvim_win_set_cursor(win, { line, 1 })
	vim.api.nvim_win_call(win, function()
		vim.cmd("normal! zv")
	end)
end

-- commenting the two lines below also don't trigger the issue
go(1)
vim.cmd([[bdelete]])

-- this will trigger the issue
-- instead of at line 100, you'll be at line roughly the same as the editor height
go(100)
```

folke commented:
Since I've now found out that that `spitkeep` issue was not caused by this PR, I've removed that specific change for now.

I'll open a new issue specifically for the `splitkeep` problem.

folke commented:
Thank you for merging!
